components/IscnUploadForm.vue

```diff
@@ -451,6 +456,106 @@ export default class UploadForm extends Vue {
     }
   }
 
+  async processEPub({ buffer, file }: { buffer: ArrayBuffer; file: File }) {
+    try {
+      const Book = ePub(buffer)
+      await Book.ready
+      const epubMetadata: any = {}
+
+      // Get metadata
+      const { metadata } = Book.packaging
+      if (metadata) {
+        epubMetadata.epubFileName = file.name
+        epubMetadata.title = metadata.title
+        epubMetadata.author = metadata.creator
+        epubMetadata.language = this.formatLanguage(metadata.language)
+        epubMetadata.description = metadata.description
+      }
+
+      // Get tags
+      const opfFilePath = await (Book.path as any).path
+      const opfContent = await Book.archive.getText(opfFilePath)
+      const parser = new DOMParser()
+      const opfDocument = parser.parseFromString(opfContent, 'application/xml')
+      const dcSubjectElements = opfDocument.querySelectorAll(
+        'dc\\:subject, subject',
+      )
+      const subjects: string[] = []
+      dcSubjectElements.forEach((element) => {
+        const subject = element.textContent
+        subject && subjects.push(subject)
+      })
+      epubMetadata.tags = subjects
+
+      // Get cover file
+      const coverUrl = (Book as any).cover
+      if (!coverUrl) {
+        this.epubMetadataList.push(epubMetadata)
+        return
+      }
+
+      const blobData = await Book.archive.getBlob(coverUrl)
+      if (blobData) {
+        const coverFile = new File([blobData], `${metadata.title}_cover.jpeg`, {
+          type: 'image/jpeg',
+        })
+
+        const fileBytes = (await fileToArrayBuffer(
+          coverFile,
+        )) as unknown as ArrayBuffer
+        if (fileBytes) {
+          const [
+            fileSHA256,
+            imageType,
+            ipfsHash,
+            // eslint-disable-next-line no-await-in-loop
+          ] = await Promise.all([
+            digestFileSHA256(fileBytes),
+            readImageType(fileBytes),
+            Hash.of(Buffer.from(fileBytes)),
+          ])
+
+          const fileRecord: any = {
+            fileName: coverFile.name,
+            fileSize: coverFile.size,
+            fileType: coverFile.type,
+            fileBlob: coverFile,
+            ipfsHash,
+            fileSHA256,
+            isFileImage: !!imageType,
+          }
+
+          epubMetadata.ipfsHash = ipfsHash
+
+          const reader = new FileReader()
+          reader.onload = (e) => {
+            if (!e.target) return
+            fileRecord.fileData = e.target.result as string
+          }
+          reader.readAsDataURL(coverFile)
+          this.epubMetadataList = [
+            ...this.epubMetadataList,
+            epubMetadata,
+          ]
+          this.fileRecords.push(fileRecord)
```

Comment:
Should wrap this `onload` function in await, dangerous use of race condition here
or a simpler way is if we don't need `fileRecord` if `fileRecord.fileData` not exists, we should put `this.fileRecords.push(fileRecord)` inside `onload`